traversal/planner/GraphPlanner.java

```diff
@@ -368,6 +279,70 @@ private void optimise(GraphManager graph, boolean singleUse) {
         printDebug(start, endSolver, end);
     }
 
+    private void updateOptimiserCoefficients() {
+        vertices.values().forEach(PlannerVertex::updateOptimiserCoefficients);
+        edges.forEach(PlannerEdge::updateOptimiserCoefficients);
+
+        if (LOG.isTraceEnabled()) LOG.trace(optimiser.toString());
+    }
+
+    private void computeTraversalCosts(GraphManager graphMgr) {
+        if (snapshot < graphMgr.data().stats().snapshot()) {
+            // TODO: we should not include the graph's uncommitted writes, but only the persisted counts in the costs
+            snapshot = graphMgr.data().stats().snapshot();
+            computeTotalCost(graphMgr);
+
+            if (!isUpToDate) {
+                recordedTotalCost = totalCost;
+                vertices.values().forEach(PlannerVertex::recordCost);
+                edges.forEach(PlannerEdge::recordCost);
+            }
+        }
+    }
+
+    private void computeTotalCost(GraphManager graphMgr) {
+        vertices.values().forEach(v -> {
+            v.computeCost(graphMgr);
+            if (costChangeSignificant(v)) setOutOfDate();
+        });
+        edges.forEach(e -> {
+            e.computeCost(graphMgr);
+            if (costChangeSignificant(e)) setOutOfDate();
+        });
+
+        double vertexCost = vertices.values().stream().mapToDouble(PlannerVertex::cost).sum();
+        double edgeCost = edges.stream().mapToDouble(PlannerEdge::cost).sum();
+        totalCost = vertexCost + edgeCost;
+        if (totalCostChangeSignificant()) setOutOfDate();
+    }
+
+    boolean costChangeSignificant(PlannerVertex<?> vertex) {
+        return costChangeSignificant(vertex.recordedCost, vertex.cost());
+    }
+
+    boolean costChangeSignificant(PlannerEdge<?, ?> edge) {
+        return costChangeSignificant(edge.forward) || costChangeSignificant(edge.backward);
+    }
+
+    boolean costChangeSignificant(PlannerEdge.Directional<?, ?> edge) {
+        return costChangeSignificant(edge.recordedCost, edge.cost());
+    }
+
+    boolean costChangeSignificant(double costPrevious, double costNext) {
+        assert recordedTotalCost > 0;
+        assert costPrevious > 0;
+        assert costNext > 0;
+
+        return (costNext / costPrevious >= OBJECTIVE_VARIABLE_COST_MAX_CHANGE ||
+                costNext / costPrevious <= 1.0 / OBJECTIVE_VARIABLE_COST_MAX_CHANGE) &&
+                abs(costNext - costPrevious) / recordedTotalCost >= OBJECTIVE_VARIABLE_TO_PLANNER_COST_MIN_CHANGE;
```

Comment:
this is actually a pretty significant fix (?) compared the previous check we had implemented? Definitely worth a comment PR if so!

Reply:
I guess? The logic is that if doubling a cost is a significant change, then so is halving.